tools/run_block.py

```diff
@@ -0,0 +1,187 @@
+from chia.types.full_block import FullBlock
+from chia.types.generator_types import BlockGenerator, GeneratorArg
+from chia.types.name_puzzle_condition import NPC
+from chia.util.config import load_config
+from chia.util.default_root import DEFAULT_ROOT_PATH
+from chia.util.errors import ConsensusError, Err
+from chia.util.ints import uint32
+from chia.wallet.cat_wallet.cat_utils import match_cat_puzzle
+
+
+@dataclass
+class CAT:
+    tail_hash: str
+    memo: str
+    npc: NPC
+
+    def cat_to_dict(self):
+        return {"tail_hash": self.tail_hash, "memo": self.memo, "npc": npc_to_dict(self.npc)}
+
+
+def condition_with_args_to_dict(condition_with_args: ConditionWithArgs):
+    return {
+        "condition_opcode": condition_with_args.opcode.name,
+        "arguments": [arg.hex() for arg in condition_with_args.vars],
+    }
+
+
+def condition_list_to_dict(condition_list: Tuple[ConditionOpcode, List[ConditionWithArgs]]):
+    assert all([condition_list[0] == cwa.opcode for cwa in condition_list[1]])
+    return [condition_with_args_to_dict(cwa) for cwa in condition_list[1]]
+
+
+def npc_to_dict(npc: NPC):
+    return {
+        "coin_name": npc.coin_name.hex(),
+        "conditions": [{"condition_type": c[0].name, "conditions": condition_list_to_dict(c)} for c in npc.conditions],
+        "puzzle_hash": npc.puzzle_hash.hex(),
+    }
+
+
+def run_generator(block_generator: BlockGenerator, constants: ConsensusConstants) -> List[CAT]:
+    npc_result = get_name_puzzle_conditions(
+        block_generator,
+        constants.MAX_BLOCK_COST_CLVM,  # min(self.constants.MAX_BLOCK_COST_CLVM, block.transactions_info.cost),
+        cost_per_byte=constants.COST_PER_BYTE,
+        mempool_mode=False,
+    )
+    if npc_result.error is not None:
+        raise ConsensusError(Err(npc_result.error))
+
+    cat_list: List[CAT] = []
+    for npc in npc_result.npc_list:
+        _, puzzle, solution = get_puzzle_and_solution_for_coin(
+            block_generator, coin_name=npc.coin_name, max_cost=constants.MAX_BLOCK_COST_CLVM
+        )
+        matched, curried_args = match_cat_puzzle(puzzle)
+
+        if matched:
+            _, tail_hash, _ = curried_args
+            memo = ""
+
+            # do somethign like this to get the memo out
+            result = puzzle.run(solution)
+            for condition in result.as_python():
+                if condition[0] == ConditionOpcode.CREATE_COIN and len(condition) >= 4:
+                    # If only 3 elements (opcode + 2 args), there is no memo, this is ph, amount
+                    if type(condition[3]) != list:
+                        # If it's not a list, it's not the correct format
+                        continue
+
+                    # special retirement address
+                    if condition[3][0].hex() == "0000000000000000000000000000000000000000000000000000000000000000":
+                        if len(condition[3]) >= 2:
+                            memo = condition[3][1].decode("utf-8")
```

Comment:
it seems there's a risk this string might not be correct UTF-8. It might make sense to catch an exception here and maybe treat it as latin-1 or something like that, rather than aborting. This is, potentially, reading any untrusted data from the chain, right?

Reply:
Added try/except here - good idead